File: DS/MinHeap.rb
class MinHeap
  attr_accessor :heap

  def initialize(array, &comparator)
    @cmp = comparator || lambda { |a, b| a < b }
    array ||= []
    @heap = build_heap(array)
  end

  def build_heap(array)
    curr_idx = (array.size - 2) / 2
    while curr_idx >= 0
      sift_down(curr_idx, array)
      curr_idx -= 1
    end
    return array
  end

  def sift_down(i = 0, array = @heap)
    curr_idx = i
    left_idx = left(curr_idx)

    while (left_idx < array.size)
      right_idx = right(curr_idx)
      right_idx = nil if (right_idx > array.length - 1)
      if right_idx != nil && @cmp.call(array[right_idx], array[left_idx])
        idx_to_swap = right_idx
      else
        idx_to_swap = left_idx
      end
      if @cmp.call(array[idx_to_swap], array[curr_idx])
        swap(curr_idx, idx_to_swap, array)
        curr_idx = idx_to_swap
        left_idx = left(curr_idx)
      else
        return
      end
    end
  end

  def sift_up(idx)
    parent_idx = parent(idx)
    curr_idx = idx
    while curr_idx > 0 && @cmp.call(@heap[curr_idx], @heap[parent_idx])
      swap(curr_idx, parent_idx)
      curr_idx = parent_idx
      parent_idx = parent(curr_idx)
    end
    return @heap
  end

  def peek
    return @heap[0] unless @heap.empty?
    return nil
  end

  def pop
    remove
  end

  def remove
    swap(0, @heap.size - 1)
    val_to_remove = @heap.pop
    sift_down()
    val_to_remove
  end

  def insert(val)
    @heap.push(val)
    sift_up(@heap.size - 1)
  end

  def swap(i, j, array = @heap)
    array[i], array[j] = array[j], array[i]
  end

  def left(idx)
    2 * idx + 1
  end

  def right(idx)
    2 * idx + 2
  end

  def parent(idx)
    (idx - 1) / 2
  end

  def size
    @heap.size
  end

  def is_empty?
    @heap.empty?
  end
end

File: DS/MinHeap.js

class MinHeap {
  constructor(array) {
    this.heap = this.buildHeap(array);
	}
	
	isEmpty() {
		return this.heap.length === 0;
	}

  buildHeap(array) {
    let currIdx = Math.floor((array.length - 2) / 2);
    while (currIdx >= 0) {
      this.siftDown(array, currIdx);
      currIdx -= 1;
    }
    return array;
  }

  siftDown(heap = this.heap, currIdx = 0) {
    let leftIdx = this.left(currIdx);
    while (leftIdx < heap.length) {
      let rightIdx = this.right(currIdx);
      if (rightIdx > heap.length - 1) rightIdx = -1;
      let idxToSwap;
      if (rightIdx !== -1 && heap[rightIdx] < heap[leftIdx]) {
        idxToSwap = rightIdx;
      } else idxToSwap = leftIdx;
      if (heap[idxToSwap] < heap[currIdx]) {
        this.swap(currIdx, idxToSwap, heap);
        currIdx = idxToSwap;
        leftIdx = this.left(currIdx);
      } else return;
    }
  }

  siftUp(idx, heap = this.heap) {
    let parentIdx = this.parent(idx);
    while (idx > 0 && heap[idx] < heap[parentIdx]) {
      this.swap(idx, parentIdx, heap);
      idx = parentIdx;
      parentIdx = this.parent(idx);
    }
    return heap;
  }

  peek() {
    if (this.heap.length) return this.heap[0];
  }

  remove(heap = this.heap) {
    this.swap(0, heap.length - 1, heap);
    const valueToRemove = heap.pop();
    this.siftDown(heap);

    return valueToRemove;
  }

  insert(value, heap = this.heap) {
    heap.push(value);
    this.siftUp(heap.length - 1, heap);
  }

  left(idx) {
    return 2 * idx + 1;
  }

  right(idx) {
    return 2 * idx + 2;
  }

  parent(idx) {
    return Math.floor((idx - 1) / 2);
  }

  swap(i, j, heap = this.heap) {
    const temp = heap[j];
    heap[j] = heap[i];
    heap[i] = temp;
  }
}

// Do not edit the line below.
exports.MinHeap = MinHeap;
